File: trbac-go/context.go
package trbac

// Context provides authorization decision information
type Context interface {
    // Action returns the operation being attempted
    Action() string
    
    // ResourceType returns the type of resource being accessed
    ResourceType() string
    
    // Roles returns all roles assigned to the actor
    Roles() []string
}

// LiteralContext is a simple Context implementation
type LiteralContext struct {
    action       string
    resourceType string
    roles        []string
}

func (lc LiteralContext) Action() string {
    return lc.action
}

func (lc LiteralContext) ResourceType() string {
    return lc.resourceType
}

func (lc LiteralContext) Roles() []string {
    return lc.roles
}

// NewLiteralContext creates a new LiteralContext
func NewLiteralContext(action, resourceType string, roles []string) Context {
    return LiteralContext{action, resourceType, roles}
}

File: trbac-haskell/src/TRBAC/Context.hs
module TRBAC.Context where

import TRBAC.Types

-- | Type class for authorization contexts
class Context c where
  -- | Get the action being performed
  getAction :: c -> Action
  
  -- | Get the type of resource being accessed
  getResourceType :: c -> ResourceType
  
  -- | Get all roles assigned to the actor
  getRoles :: c -> [Role]

-- | A simple context implementation
data BasicContext = BasicContext
  { basicAction :: Action
  , basicResourceType :: ResourceType
  , basicRoles :: [Role]
  }

instance Context BasicContext where
  getAction = basicAction
  getResourceType = basicResourceType
  getRoles = basicRoles